test: add P2PK support to MiniWallet

theStack opened the pull request:
This PR adds support for creating and spending transactions with raw pubkey (P2PK) outputs to MiniWallet, [as suggested by MarcoFalke](https://github.com/bitcoin/bitcoin/pull/21900#discussion_r629524841). Using that  mode in the test `feature_csv_activation.py`, all txs submitted to the mempool follow the standard policy, i.e. `-acceptnonstdtxn=1` can be removed.

Possible follow-ups:
* Improve MiniWallet constructor Interface; an enum-like parameter instead of two booleans would probably be better
* Look at other tests that could benefit from P2PK (e.g. feature_cltv.py?)
* Check vsize also for P2PK txs (vsize varies due to signature, i.e. a range has to be asserted)

theStack committed:
test: MiniWallet: add P2PK support

theStack committed:
test: use P2PK-MiniWallet for feature_csv_activation.py

Using the MiniWallet in P2PK mode, all transactions submitted to the
mempool are following the standard policy now, i.e. the node command
line parameter '-acceptnonstdtxn=1' is not needed anymore.

reemuru commented on test/functional/feature_csv_activation.py:
```diff
@@ -103,12 +102,14 @@ def create_self_transfer_from_utxo(self, input_tx):
     def create_bip112special(self, input, txversion):
         tx = self.create_self_transfer_from_utxo(input)
```
Hey, I'm hoping to be new contributor someday. Just taking a peek at some PRs. Could these original tx assignments be removed since it is being re-assigned?

```cpp
tx = self.create_self_transfer_from_utxo(input)
```

theStack replied:
Hi reemuru, thanks for taking a look and welcome as new contributor!
Note that the `tx` instance created in this line is used on the right-hand-side of the re-assignment two lines below, i.e. it serves as input for the `sign_tx` method call: `tx = self.miniwallet.sign_tx(tx)`. The best would probably be if `sign_tx` modifies the tx in-place and doesn't return anything, leading to simply `self.miniwallet.sign_tx(tx)` to have cleaner code and reduce confusion. Will check if that is possible.

MarcoFalke commented on test/functional/feature_csv_activation.py:
```diff
@@ -90,7 +90,6 @@ def set_test_params(self):
         self.setup_clean_chain = True
         self.extra_args = [[
             '-whitelist=noban@127.0.0.1',
-            '-acceptnonstdtxn=1',
```
Nice. Thanks for working on this.

I guess an alternative to consider might be to use the option to ignore specific reject reasons, when it is availalbe. #20753

MarcoFalke commented:
> Look at other tests that could benefit from P2PK (e.g. feature_cltv.py?)

I think that test also uses the mempool, which doesn't like raw scripts. I guess an alternative to that is to wrap all scripts into a hashed output first, which is cumbersome :(. Or ignore specific reject reasons, when available.

Modusto25 commented:
What do you need me to do ?()

On Mon, May 17, 2021, 12:20 AM MarcoFalke ***@***.***> wrote:

> Look at other tests that could benefit from P2PK (e.g. feature_cltv.py?)
>
> I think that test also uses the mempool, which doesn't like raw scripts. I
> guess an alternative to that is to wrap all scripts into a hashed output
> first, which is cumbersome :(. Or ignore specific reject reasons, when
> available.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/21945#issuecomment-842074149>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AUA7ISEVITP5QD3GGUJBW3DTOC7VFANCNFSM44265U2A>
> .
>

michaelfolkson commented:
Concept ACK. A clear improvement to be able to generate P2PK outputs in the functional tests in addition to anyone-can-spend outputs.

theStack commented:
Thanks for the reviews! #20753 looks interesting, though at first glance it seems to be discussed controversially. Will look at it more in detail within the next days. Another example where the MiniWallet with P2PK support helps is `feature_dersig.py`, which used the wallet for creating a DER-signature (i.e. ignoring specific reject reasons doesn't help in this case).

MarcoFalke commented on test/functional/test_framework/wallet.py:
```diff
@@ -27,12 +31,20 @@
+                self._scriptPubKey = bytes(CScript([pub_key.get_bytes(), OP_CHECKSIG]))
+            else:
+                self._scriptPubKey = bytes(CScript([OP_TRUE]))
         else:
```
would be nice to make this an `elif use_p2pk` to make it easier to switch to an enum

theStack replied:
Thanks, done.

theStack commented:
Force-pushed with the following changes:
* changed `MiniWallet.sign_tx()` method to not return the tx again, as the sign works in-place by directly modifying the first input's `scriptSig` (thanks to @reemuru whose [review comment](https://github.com/bitcoin/bitcoin/pull/21945#discussion_r633181315) inspired me to do this!)
* changed mode selection structure to be more friendlier w.r.t. changing to an enum (thanks to @MarcoFalke for [suggesting](https://github.com/bitcoin/bitcoin/pull/21945#discussion_r633706544)). The mode selection looks now like the following:

| raw_script  | use_p2pk | Mode |
| ------------- | ------------- | ---------- |
| True  | X  | raw OP_TRUE output |
| X  | True  | raw P2PK (<pubkey> OP_CHECKSIG) output |
| False | False | bech32 P2WSH address, containing hash of OP_TRUE (default) |

laanwj commented:
Code review ACK 4bea30169218e2f21e0c93a059966b41c8edd205
